File: Practice/Greedy/q06/q06.py
# 무지의 먹방 라이브 317p 513p
import heapq

def solution(food_times, k):
    if sum(food_times) <= k: return -1

    q = []
    for idx in range(len(food_times)):
        heapq.heappush(q, (food_times[idx], idx+1))

    used_time = 0
    prev_time = 0
    left_food = len(food_times)
    while used_time + ((q[0][0] - prev_time) * left_food) <= k:
        now = heapq.heappop(q)[0]
        used_time += (now - prev_time) * left_food
        prev_time = now
        left_food -= 1
    
    result = sorted(q, key=lambda x:x[1])
    return result[(k-used_time) % left_food][1]

File: Practice/Greedy/q06/q06.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <queue>

using namespace std;

priority_queue<pair<int,int>> food;

bool compare(pair<int,int> a, pair<int,int> b){
    return a.second < b.second;
}

int solution(vector<int> food_times, long long k) {
    
    long long sum_time = 0;
    for(int i = 0; i < food_times.size(); i++){
        sum_time += food_times[i];
    }
    if(sum_time <= k) return -1;
    
    for(int i = 0; i < food_times.size(); i++){
        food.push({-food_times[i], i+1});
    }

    sum_time = 0;
    long long prev_time = 0;
    long long left_food = food_times.size();
    while(sum_time+((-food.top().first-prev_time)*left_food) <= k){
        int now = -food.top().first;
        sum_time += (now-prev_time)*left_food;
        prev_time = now;
        left_food--;
        food.pop();
    }
    
    vector<pair<int,int>> result;
    while(!food.empty()){
        int t = -food.top().first;
        int n = food.top().second;
        result.push_back({t, n});
        food.pop();
    }
    sort(result.begin(), result.end(), compare);
    return result[(k-sum_time) % left_food].second;
}

int main(void){
    printf("%d\n",solution({3, 1, 2}, 5));
}